File: Tests/ReminderCoreTests/PresentationRoutingTests.swift
import XCTest
@testable import ReminderCore

/// A presenter that records how each reminder would be shown, standing in for
/// the app's real routing (subtle card / notification / full-screen overlay).
///
/// The app layer cannot be imported here, so this mirrors the same decision the
/// `OverlayPresenter` makes. The point is to pin down the routing rules, which
/// are the part a user actually feels.
final class RoutingPresenter: ReminderPresenting {
    enum Surface: Equatable {
        case subtleCard
        case notification
        case fullScreenOverlay
    }

    private(set) var routed: [(title: String, surface: Surface)] = []
    private(set) var dismissAllCount = 0

    /// Mirrors `OverlayPresenter.present`.
    static func surface(for priority: Priority) -> Surface {
        switch priority {
        case .subtle: return .subtleCard
        case .normal, .important: return .notification
        case .critical: return .fullScreenOverlay
        }
    }

    func present(_ reminder: Reminder, settings: Settings) {
        routed.append((reminder.title, Self.surface(for: reminder.priority)))
    }

    func dismissAll() { dismissAllCount += 1 }
}

@MainActor
final class PresentationRoutingTests: XCTestCase {

    private var calendar: Calendar = {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        return cal
    }()

    private func makeDueReminder(
        title: String, priority: Priority, at start: Date
    ) -> Reminder {
        var reminder = Reminder(
            title: title, schedule: .interval(minutes: 10),
            priority: priority, createdAt: start
        )
        reminder.lastFiredAt = start
        return reminder
    }

    /// Each tier must reach a genuinely different surface. This is the whole
    /// premise of the app: a 20-minute nudge and an hourly pressure-relief
    /// prompt must not feel the same.
    func testEachPriorityRoutesToItsOwnSurface() {
        XCTAssertEqual(RoutingPresenter.surface(for: .subtle), .subtleCard)
        XCTAssertEqual(RoutingPresenter.surface(for: .normal), .notification)
        XCTAssertEqual(RoutingPresenter.surface(for: .important), .notification)
        XCTAssertEqual(RoutingPresenter.surface(for: .critical), .fullScreenOverlay)
    }

    func testFiredRemindersAreRoutedByPriority() {
        let start = Date(timeIntervalSince1970: 1_800_000_000)
        let reminders = [
            makeDueReminder(title: "Shift", priority: .subtle, at: start),
            makeDueReminder(title: "Water", priority: .normal, at: start),
            makeDueReminder(title: "Meds", priority: .important, at: start),
            makeDueReminder(title: "Tilt", priority: .critical, at: start),
        ]
        let store = InMemoryDataStore(data: AppData(reminders: reminders))
        let clock = MutableDateProvider(now: start)
        let presenter = RoutingPresenter()
        let engine = ReminderEngine(
            store: store, dateProvider: clock, presenter: presenter, calendar: calendar
        )

        clock.advance(by: 10 * 60)
        engine.tick()

        XCTAssertEqual(
            presenter.routed.map(\.surface),
            [.subtleCard, .notification, .notification, .fullScreenOverlay],
            "Presented lowest priority first, so the overlay lands on top"
        )
    }

    /// The default reminders must map to the presentations they were designed
    /// for: the hourly tilt takes over the screen, the 20-minute shift does not.
    func testStarterSetRoutesTheWayItWasDesignedTo() {
        let starters = DefaultReminders.starterSet()

        let tilt = starters.first { $0.title == "Tilt Back" }!
        XCTAssertEqual(
            RoutingPresenter.surface(for: tilt.priority), .fullScreenOverlay,
            "The hourly pressure-relief prompt must interrupt"
        )

        let shift = starters.first { $0.title == "Weight Shift" }!
        XCTAssertEqual(
            RoutingPresenter.surface(for: shift.priority), .subtleCard,
            "A nudge every 20 minutes must stay quiet or it gets tuned out"
        )

        let water = starters.first { $0.title == "Drink Water" }!
        XCTAssertEqual(RoutingPresenter.surface(for: water.priority), .notification)
    }

    /// Pausing must clear anything already on screen, not just stop new ones.
    func testPausingClearsWhateverIsOnScreen() {
        let start = Date(timeIntervalSince1970: 1_800_000_000)
        let reminder = makeDueReminder(title: "Tilt", priority: .critical, at: start)
        let store = InMemoryDataStore(data: AppData(reminders: [reminder]))
        let clock = MutableDateProvider(now: start)
        let presenter = RoutingPresenter()
        let engine = ReminderEngine(
            store: store, dateProvider: clock, presenter: presenter, calendar: calendar
        )

        clock.advance(by: 10 * 60)
        engine.tick()
        XCTAssertEqual(presenter.routed.count, 1)

        engine.setPaused(true)
        XCTAssertEqual(presenter.dismissAllCount, 1)
    }

    /// A reminder with an activity duration drives the countdown, so the value
    /// has to survive into what gets presented.
    func testActivityDurationReachesThePresenter() {
        let start = Date(timeIntervalSince1970: 1_800_000_000)
        var reminder = makeDueReminder(title: "Tilt", priority: .critical, at: start)
        reminder.activityDurationSeconds = 5 * 60

        let store = InMemoryDataStore(data: AppData(reminders: [reminder]))
        let clock = MutableDateProvider(now: start)
        let presenter = RecordingPresenter()
        let engine = ReminderEngine(
            store: store, dateProvider: clock, presenter: presenter, calendar: calendar
        )

        clock.advance(by: 10 * 60)
        engine.tick()

        XCTAssertEqual(presenter.presented.first?.activityDurationSeconds, 300)
    }
}

File: Windows/Pauselet.Core.Tests/PresentationRoutingTests.cs
using NodaTime;
using Pauselet.Core;
using Xunit;

namespace Pauselet.Core.Tests;

/// <summary>
/// A presenter that records how each reminder would be shown, standing in for
/// the app's real routing (subtle card / notification / full-screen overlay).
///
/// The app layer cannot be imported here, so this mirrors the same decision the
/// real presenter makes. The point is to pin down the routing rules, which are
/// the part a user actually feels.
/// </summary>
public sealed class RoutingPresenter : IReminderPresenting
{
    public enum Surface
    {
        SubtleCard,
        Notification,
        FullScreenOverlay,
    }

    public List<(string Title, Surface Surface)> Routed { get; } = [];
    public int DismissAllCount { get; private set; }

    /// <summary>Mirrors the real presenter's routing decision.</summary>
    public static Surface SurfaceFor(Priority priority) => priority switch
    {
        Priority.Subtle => Surface.SubtleCard,
        Priority.Normal or Priority.Important => Surface.Notification,
        Priority.Critical => Surface.FullScreenOverlay,
        _ => throw new ArgumentOutOfRangeException(nameof(priority)),
    };

    public void Present(Reminder reminder, Settings settings)
    {
        Routed.Add((reminder.Title, SurfaceFor(reminder.Priority)));
    }

    public void DismissAll() => DismissAllCount += 1;
}

public class PresentationRoutingTests
{
    private static readonly DateTimeZone Utc = DateTimeZone.Utc;

    private static Reminder MakeDueReminder(string title, Priority priority, Instant start) =>
        new Reminder
        {
            Title = title, Schedule = new Schedule.Interval(10),
            Priority = priority, CreatedAt = start,
        } with { LastFiredAt = start };

    /// <summary>
    /// Each tier must reach a genuinely different surface. This is the whole
    /// premise of the app: a 20-minute nudge and an hourly pressure-relief
    /// prompt must not feel the same.
    /// </summary>
    [Fact]
    public void EachPriorityRoutesToItsOwnSurface()
    {
        Assert.Equal(RoutingPresenter.Surface.SubtleCard,
            RoutingPresenter.SurfaceFor(Priority.Subtle));
        Assert.Equal(RoutingPresenter.Surface.Notification,
            RoutingPresenter.SurfaceFor(Priority.Normal));
        Assert.Equal(RoutingPresenter.Surface.Notification,
            RoutingPresenter.SurfaceFor(Priority.Important));
        Assert.Equal(RoutingPresenter.Surface.FullScreenOverlay,
            RoutingPresenter.SurfaceFor(Priority.Critical));
    }

    [Fact]
    public void FiredRemindersAreRoutedByPriority()
    {
        var start = Instant.FromUnixTimeSeconds(1_800_000_000);
        var reminders = new[]
        {
            MakeDueReminder("Shift", Priority.Subtle, start),
            MakeDueReminder("Water", Priority.Normal, start),
            MakeDueReminder("Meds", Priority.Important, start),
            MakeDueReminder("Tilt", Priority.Critical, start),
        };
        var store = new InMemoryDataStore(new AppData { Reminders = reminders });
        var clock = new MutableDateProvider(start);
        var presenter = new RoutingPresenter();
        var engine = new ReminderEngine(store, clock, presenter, Utc);

        clock.AdvanceSeconds(10 * 60);
        engine.Tick();

        // Presented lowest priority first, so the overlay lands on top.
        Assert.Equal(
            [
                RoutingPresenter.Surface.SubtleCard,
                RoutingPresenter.Surface.Notification,
                RoutingPresenter.Surface.Notification,
                RoutingPresenter.Surface.FullScreenOverlay,
            ],
            presenter.Routed.Select(r => r.Surface).ToArray()
        );
    }

    /// <summary>
    /// The default reminders must map to the presentations they were designed
    /// for: the hourly tilt takes over the screen, the 20-minute shift does
    /// not.
    /// </summary>
    [Fact]
    public void StarterSetRoutesTheWayItWasDesignedTo()
    {
        var starters = DefaultReminders.StarterSet();

        var tilt = starters.First(r => r.Title == "Tilt Back");
        // The hourly pressure-relief prompt must interrupt.
        Assert.Equal(
            RoutingPresenter.Surface.FullScreenOverlay,
            RoutingPresenter.SurfaceFor(tilt.Priority)
        );

        var shift = starters.First(r => r.Title == "Weight Shift");
        // A nudge every 20 minutes must stay quiet or it gets tuned out.
        Assert.Equal(
            RoutingPresenter.Surface.SubtleCard,
            RoutingPresenter.SurfaceFor(shift.Priority)
        );

        var water = starters.First(r => r.Title == "Drink Water");
        Assert.Equal(
            RoutingPresenter.Surface.Notification,
            RoutingPresenter.SurfaceFor(water.Priority)
        );
    }

    /// <summary>Pausing must clear anything already on screen, not just stop new ones.</summary>
    [Fact]
    public void PausingClearsWhateverIsOnScreen()
    {
        var start = Instant.FromUnixTimeSeconds(1_800_000_000);
        var reminder = MakeDueReminder("Tilt", Priority.Critical, start);
        var store = new InMemoryDataStore(new AppData { Reminders = [reminder] });
        var clock = new MutableDateProvider(start);
        var presenter = new RoutingPresenter();
        var engine = new ReminderEngine(store, clock, presenter, Utc);

        clock.AdvanceSeconds(10 * 60);
        engine.Tick();
        Assert.Single(presenter.Routed);

        engine.SetPaused(true);
        Assert.Equal(1, presenter.DismissAllCount);
    }

    /// <summary>
    /// A reminder with an activity duration drives the countdown, so the value
    /// has to survive into what gets presented.
    /// </summary>
    [Fact]
    public void ActivityDurationReachesThePresenter()
    {
        var start = Instant.FromUnixTimeSeconds(1_800_000_000);
        var reminder = MakeDueReminder("Tilt", Priority.Critical, start) with
        {
            ActivityDurationSeconds = 5 * 60,
        };

        var store = new InMemoryDataStore(new AppData { Reminders = [reminder] });
        var clock = new MutableDateProvider(start);
        var presenter = new RecordingPresenter();
        var engine = new ReminderEngine(store, clock, presenter, Utc);

        clock.AdvanceSeconds(10 * 60);
        engine.Tick();

        Assert.Equal(300, presenter.Presented.First().ActivityDurationSeconds);
    }
}

/// <summary>Tests for the per-reminder on-screen duration.</summary>
public class DisplayDurationTests
{
    /// <summary>Mirrors how the presenter picks a duration for a subtle card.</summary>
    private static int EffectiveSeconds(Reminder reminder, Settings settings) =>
        Math.Max(2, reminder.DisplaySeconds ?? settings.SubtleDisplaySeconds);

    [Fact]
    public void FallsBackToTheGlobalSettingWhenUnset()
    {
        var settings = new Settings { SubtleDisplaySeconds = 8 };
        var reminder = new Reminder { Title = "Shift", Schedule = new Schedule.Interval(20) };

        Assert.Null(reminder.DisplaySeconds);
        Assert.Equal(8, EffectiveSeconds(reminder, settings));
    }

    /// <summary>The point of the feature: a wordy reminder can be given longer to read.</summary>
    [Fact]
    public void PerReminderValueOverridesTheGlobalSetting()
    {
        var settings = new Settings { SubtleDisplaySeconds = 8 };
        var reminder = new Reminder
        {
            Title = "Shift", Schedule = new Schedule.Interval(20), DisplaySeconds = 25,
        };

        Assert.Equal(25, EffectiveSeconds(reminder, settings));
    }

    /// <summary>A too-short value would flash the card and vanish before it can be read.</summary>
    [Fact]
    public void DurationIsClampedToAReadableMinimum()
    {
        var settings = new Settings { SubtleDisplaySeconds = 8 };
        var reminder = new Reminder
        {
            Title = "Shift", Schedule = new Schedule.Interval(20), DisplaySeconds = 0,
        };

        Assert.Equal(2, EffectiveSeconds(reminder, settings));
    }

    [Fact]
    public void DisplaySecondsSurvivesARoundTrip()
    {
        var directory = Path.Combine(Path.GetTempPath(), $"ReminderDisplay-{Guid.NewGuid()}");
        Directory.CreateDirectory(directory);
        try
        {
            var store = new FileDataStore(Path.Combine(directory, "data.json"));
            var reminder = new Reminder
            {
                Title = "Shift", Schedule = new Schedule.Interval(20), DisplaySeconds = 30,
            };

            store.Save(new AppData { Reminders = [reminder] });
            var loaded = store.Load();

            Assert.Equal(30, loaded.Reminders.First().DisplaySeconds);
        }
        finally
        {
            Directory.Delete(directory, recursive: true);
        }
    }

    /// <summary>Reminders saved before this field existed must still load.</summary>
    [Fact]
    public void OlderDataWithoutDisplaySecondsStillLoads()
    {
        var directory = Path.Combine(Path.GetTempPath(), $"ReminderLegacy-{Guid.NewGuid()}");
        Directory.CreateDirectory(directory);
        try
        {
            var url = Path.Combine(directory, "data.json");

            // A record written before `displaySeconds` was added.
            var legacy = $$"""
            {
              "schemaVersion": 1,
              "reminders": [{
                "id": "{{Guid.NewGuid().ToString().ToUpperInvariant()}}",
                "title": "Weight Shift",
                "message": "Shift your weight.",
                "schedule": { "interval": { "minutes": 20 } },
                "priority": "subtle",
                "isEnabled": true,
                "symbolName": "arrow.up.and.down.circle",
                "createdAt": 1780000000
              }],
              "settings": {
                "quietHours": {
                  "isEnabled": false, "startHour": 22, "startMinute": 0,
                  "endHour": 7, "endMinute": 0, "allowsCritical": true
                },
                "isPaused": false, "snoozeMinutes": 5, "subtleDisplaySeconds": 8,
                "launchAtLogin": false, "showsNextReminderInMenuBar": true,
                "soundEnabled": true
              },
              "events": []
            }
            """;
            File.WriteAllText(url, legacy);

            var loaded = new FileDataStore(url).Load();
            Assert.Single(loaded.Reminders);
            // A missing field should decode as null, not fail the whole file.
            Assert.Null(loaded.Reminders.First().DisplaySeconds);
        }
        finally
        {
            Directory.Delete(directory, recursive: true);
        }
    }
}
